File: src/leetcode/Stack/Binary_Tree_Preorder_Traversal_144.java
package leetcode.Stack;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

/**
 * Created by sunbo_000 on 10/19/2016.
 */
/*
    https://leetcode.com/problems/binary-tree-preorder-traversal/
    Given a binary tree, return the preorder traversal of its nodes' values.

    For example:
    Given binary tree {1,#,2,3},
       1
        \
         2
        /
       3
    return [1,2,3].

    Note: Recursive solution is trivial, could you do it iteratively?
 */

/**
 * Use Stack: root->right->left
 */
public class Binary_Tree_Preorder_Traversal_144 {

    private static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> ret = new ArrayList<>();
        if (root == null) return ret;
        LinkedList<TreeNode> stack = new LinkedList<>();
        stack.push(root);
        while (!stack.isEmpty()) {
            TreeNode node = stack.pop();
            ret.add(node.val);
            if (node.right != null)
                stack.push(node.right);
            if (node.left != null)
                stack.push(node.left);
        }

        return ret;
    }

    public static void main(String[] args) {
        TreeNode one = new TreeNode(1);
        TreeNode two = new TreeNode(2);
        TreeNode three = new TreeNode(3);
        three.right = two;
        three.left = one;
        List<Integer> result = new Binary_Tree_Preorder_Traversal_144().preorderTraversal(three);
        String stop = "";
    }
}

File: src/leetcode_javascript/Binary_Tree_Preorder_Traversal_144.js
/**
 * 144 Binary Tree Preorder Traversal
 *
 * Given a binary tree, return the preorder traversal of its nodes' values.

  For example:
  Given binary tree {1,#,2,3},
     1
      \
       2
      /
     3
  return [1,2,3].
 *
 * https://leetcode.com/problems/binary-tree-preorder-traversal/
 */

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var preorderTraversal = function(root) {
    var result = [];
    if (root === null) {
        return result;
    }
    var stack = [];
    stack.push(root);
    while (stack.length > 0) {
        var curr = stack.pop();
        result.push(curr.val);
        if (curr.right !== null) {
            stack.push(curr.right);
        }
        if (curr.left !== null) {
            stack.push(curr.left);
        }
    }
    return result;
};